crates/sui-framework/sources/governance/voting_power.move

```diff
@@ -0,0 +1,256 @@
+// Copyright (c) Mysten Labs, Inc.
+// SPDX-License-Identifier: Apache-2.0
+
+module sui::voting_power {
+    use sui::validator::{Self, Validator};
+    use std::vector;
+
+    /// Set total_voting_power as 10_000 by convention. Individual voting powers can be interpreted
+    /// as easily understandable basis points (e.g., voting_power: 100 = 1%, voting_power: 1 = 0.01%) rather than
+    /// opaque quantities whose meaning changes from epoch to epoch as the total amount staked shifts.
+    /// Fixing the total voting power allows clients to hardcode the quorum threshold and total_voting power rather
+    /// than recomputing these.
+    // TODO: we can go bigger if more precision is needed (e.g., will validators have <0.01% voting power given min stake requirements)?,
+    // but I think using a power of 10 is a useful convention.
+    const TOTAL_VOTING_POWER: u64 = 10_000;
+
+    /// Quorum threshold for our fixed voting power--any message signed by this much voting power can be trusted
+    /// up to BFT assumotions
+    const QUORUM_THRESHOLD: u64 = 6_667;
+
+    /// Cap voting power of an individual validator at 10%.
+    // TODO: determine what this should be
+    const MAX_VOTING_POWER: u64 = 1_000;
+
+    /// We should never observe this, modulo bugs
+    const EInternalInvariantViolation: u64 = 0;
+
+    /// Convert each validator's stake to a voting power normalized w.r.t `TOTAL_VOTING_POWER`,
+    /// and update `active_validators` accordingly, and attempt cap each validator's voting power at `MAX_VOTING_POWER`.
+    /// Capping is handled by redistributing the voting power "taken away from" capped validators *proportionally* among validators with voting
+    /// power less than the max.
+    /// Similarly, "leftover" voting power due to rounding error is distributed *equally* among validators with voting power less than the max (if possible),
+    /// and among all validators if everyone is already at the max.
+    /// This function ensures the following invariants:
```

Comment:
How to reason about the eventual termination of this algorithm?

Reply:
Added some comments here.